### PR TITLE
Callback Metric Dict getting overwritten by Log and Progress Bar Dict

### LINKED ISSUE #1727: Progress bar \ log dict items added to outputs in training_epoch_end
## 🐛 Bug

When running with training_step returning something along this:
```
return { 'loss':x,
            'some_value_for_epoch_end': y,
            'progress_bar':{'v1':z.mean()}}

```
you get 'v1' as part of the outputs in epoch end. This is unexpected imo.

Also in case it is:
```
return { 'loss':x,
            'some_value_for_epoch_end': y,
            'progress_bar':{'some_value_for_epoch_end': z.mean()}}

```

you get 'some_value_for_epoch_end' = z.mean(), as the values get overwritten.

It originate from file trainer/logging.py
lines 172, 173, where you overwrite the values using progress_bar_outputs + log_outputs

### PR DESCRIPTION
**Pull request by olineumann**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)

- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?

- [x] Did you make sure to update the docs?
   
- [x] Did you write any new necessary tests?
  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## What does this PR do?
Fixes #1727.

Dict values passed to progress bar or log overwriting callback values. See example in issue.

There are several options to solve it. This simply removes adding progress bar and log values to callback dict. Tests passed on my machine.

**But this will affect users code e.g. when log metric as early stopping metric was used**

## PR review    
**Opinions, other solutions, recommendations, ... welcome!** Also help in updating docs.

## Did you have fun?
🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
@olineumann Thanks for the PR. Is it correct that the bug only exists for training_epoch end, not for the valid/test_epoch_end? In this case, could you check that your change brings it in line with validation_step/epoch_end. 

> But this will affect users code e.g. when log metric as early stopping metric was used

I think consensus is that we want to do early stopping only on validation metrics, and no more on training metrics as it is right now. #1458 is dealing with this.

**Comment by olineumann** [comment]
@awaelchli No it should effect test, validation and test epoch end. See my changes on validation epoch end of the base model in the tests.

File: tests/base/model_valid_epoch_ends.py [hunk 1]
```diff
@@ -43,5 +43,7 @@ def _mean(res, key):
             val_acc_mean /= len(outputs)
 
         metrics_dict = {'val_loss': val_loss_mean.item(), 'val_acc': val_acc_mean.item()}
-        results = {'progress_bar': metrics_dict, 'log': metrics_dict}
-        return results
+        result = metrics_dict.copy()
```
Comment (Borda) [inline tests/base/model_valid_epoch_ends.py:46]: why the copy here?

Comment (olineumann) [reply tests/base/model_valid_epoch_ends.py:46]: Without copying the metric dict, result and metric dict reference the same object. So adding metric dict to result['progress_bar'] would also change metric_dict. If then adding metric dict to result['log'], result['log']['progress_bar'] would exist and cause errors in tests in my machine. 

First I reused the metric_dict by
```
metric_dict['progress_bar'] = metric_dict
metric_dict['log'] = metric_dict
return metric_dict
```
But this is wrong and leads to the same error.

File: pytorch_lightning/trainer/logging.py [hunk 2]
```diff
@@ -168,10 +168,6 @@ def process_output(self, output, train=False):
         # ---------------
         hiddens = output.get('hiddens')
 
-        # use every metric passed in as a candidate for callback
-        callback_metrics.update(progress_bar_metrics)
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/logging.py:172]: why do we need to remove this?
without this log metrics and progress bar metrics won't be candidates for the callbacks

Comment (olineumann) [reply pytorch_lightning/trainer/logging.py:172]: In #1727 @kessido had the issue, that progress bar or log metric overwrites the callback metric of the top layer dict. An example was also given by @kessido see [COLAB](https://colab.research.google.com/drive/1kYjEH2h5Ki9ygZ0XlA6hpbcxA1L1NNzb?usp=sharing)

I don't know if this needs to be fixed, that's why I asked in the issue for more opinions. Only @awaelchli responded and said he thinks that this also needs to be fixed.

Because no one started a PR I did to initiate a discussion. I have several ideas on how this could be fixed and mentioned some in the issue above. But this was the easiest and quickest solution. I didn't want to spend too much afford on a solution which then will be discarded.

**Comment by Borda** [comment]
@olineumann mind check last comments? it would be great to get this done :rabbit:

**Comment by olineumann** [comment]
> @olineumann mind check last comments? it would be great to get this done :rabbit: 

Hey Borda,

thanks for replying.

I responded to the last comments on the code reviews. I still not sure what the best way would be to solve the problem. Because the current fix would affect many users which will lead to many issues I think from users complaining their logging or early stopping won't work anymore.

I could implement that only metric values from progress bar or logging would be written to the top dict if the key didn't exists already. That wouldn't affect so much users I think.

I hoped that there were more opinions on that. I could implement the solution above, rebase and push so it could be merged.

**Comment by olineumann** [comment]
@Borda Just rebased to master, implemented and pushed the solution, and all tests passing :beers:.

Now the logging and progress bar metric values are only written to the top-level callback metric dict if the key didn't exist. Also at first, the logging values are written, and then the progress bar values (so logging metric values have a higher priority if both containing the same key). This shouldn't affect other users' code as long as they didn't use the same key in different metric dicts.

**Comment by Borda** [comment]
just thinking that it may be also solved by #1989 what do you think?

**Comment by olineumann** [comment]
> just thinking that it may be also solved by #1989 what do you think?

Didn't saw the PR before. Currently, I have not so much time to follow pytorch_lightning... But looks like a nice new feature!

I think when using the new way by passing a Result() object it is already solved. But currently the PR isn't done yet so the old way will still be used and also, as far I understand, the old way should still be supported. So I think this PR could be merged into master to fix #1727 (which wouldn't be fixed by #1989 as long the user switches to the new result object).

File: pytorch_lightning/trainer/logging.py (hunk 2 above)
Comment (sykrn) [reply pytorch_lightning/trainer/logging.py:172]: @williamFalcon @olineumann, in the current update, when that line is removed and we use Result Obj, we cannot save the model checkpoint in form of  `{val_loss}`, it will result `epoch=1-val_loss=0` which cannot get the `val_loss` due to the filename params based on the `callback_metrics`. Is there another way to assign Result/TrainResul/EvalResult Obj with callback_metrics.

**Comment by Borda** [comment]
@olineumann how is it going? can we finish it soon...

**Comment by williamFalcon** [comment]
this was solved in the structured results refactors
